lib/web/usertasks.go

```diff
@@ -113,18 +113,32 @@ func (h *Handler) userTaskListByIntegration(w http.ResponseWriter, r *http.Reque
 		return nil, trace.BadParameter("integration query param is required")
 	}
 
-	userTasks, nextKey, err := clt.UserTasksServiceClient().ListUserTasksByIntegration(r.Context(), int64(limit), startKey, integrationName)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-
-	items := make([]ui.UserTask, 0, len(userTasks))
-	for _, userTask := range userTasks {
-		items = append(items, ui.MakeUserTask(userTask))
+	taskStateFilter := values.Get("state")
+
+	items := make([]ui.UserTask, 0, limit)
+	pageLimit := limit
+	for {
+		userTasks, nextKey, err := clt.UserTasksServiceClient().ListUserTasksByIntegration(r.Context(), int64(pageLimit), startKey, integrationName)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		for _, userTask := range userTasks {
+			if taskStateFilter != "" && userTask.Spec.State != taskStateFilter {
+				continue
+			}
+			items = append(items, ui.MakeUserTask(userTask))
+		}
+
+		startKey = nextKey
+		pageLimit = limit - int32(len(items))
+		if startKey == "" || len(items) >= int(limit) {
+			break
+		}
```

Comment:
This seems like it could be a bit inefficient

If the page limit is 10, and there's a filter specified that will eventually match only 9 items in the entire database, and 1000 entries in the database - if the first request gets those 9 items, it would then fetch the remaining 990 items one by one

Reply:
Thank you.
I added a new param for listing user tasks: filters.
It accepts Integration and State as possible filters.

I marked the ListUserTasksByIntegration as deprecated, to favor the usage of the ListUserTasks, simplifying the API.